### DISCUSSION TITLE
How to FIX: has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement ?

### DISCUSSION THREAD

**Discussion by ynhhoJ** [rwf2/Rocket#2042] [Questions]
Hi!

How to use serde JSON life-time argument with rocket_sync_db_pools?

```rust
// struct
use rocket::serde::Deserialize;

#[derive(Deserialize)]
#[serde(crate = "rocket::serde")]
pub struct PostUserData<'a> {
    pub login: &'a str,
    pub password: &'a str,
}
```

```rust
// main
use rocket_sync_db_pools::{database, diesel};

#[database("db")]
pub struct DbConn(diesel::PgConnection);

#[rocket::main]
async fn main() -> Result<(), Error> {
    rocket::build()
        .attach(DbConn::fairing())
        .mount(
            "/",
            routes![
                // users
                authenticate_user,
            ],
        )
        .mount("/", rocket_cors::catch_all_options_routes())
        .manage(cors_options().to_cors().expect("To not fail"))
        .launch()
        .await
}
```

```rust
#[post(
    "/users/authenticate",
    format = "json",
    data = "<post_user_data>"
)]
pub async fn authenticate_user(
    cors: Guard<'_>,
   // `post_user_data` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
   // ...is captured here...rustc E0759
    post_user_data: Json<PostUserData<'_>>,
    db: DbConn,
) -> rocket_cors::Responder<Value> {
    use schema::users::dsl::*;

    let user_db_result = db.run(move |c| find_user(&c, post_user_data.login)).await;
    // code...
}
```

If i remove life-time argument from `post_user_data` (ex: `post_user_data: Json<PostUserData>`) everything is okay.

**Comment by somehowchris** [comment]
I am clearly not an expert on serde things and not as dedicated to using `&str` with user provided Data as it anyway would need to be allocated at runtime.

Now. I would go with `String` types instead of `&‘a str`, as said, they anyway would need to be allocated on request.

Still curious, why would you use &‘a str in user provided data?

**Reply by ynhhoJ** [reply]
`&'a str` - it's a [life-time argument](https://doc.rust-lang.org/book/ch10-03-lifetime-syntax.html). I'm trying to learn using correct types for each case in part 🤔

`PostUserData` - lifes only in `authenticate_user`, after when `authenticate_user` was executed, memory are removed

**Reply by somehowchris** [reply]
Yea, that‘s exactly why you do not know the size of `str` at compile-time if you rely on user provided data and if you were to use life time symbols you would need to go all in on your example and not only part of it

**Reply by ynhhoJ** [reply]
Thank you for response!

**Reply by ynhhoJ** [reply]
@somehowchris , where is an another question. What if i have a life-time argument in my struct which i should use witho rocket?
I think, question is not only about `'a &str` 🤔

Removing life-time args would by temporary solution for me in some concret cases

**Reply by somehowchris** [reply]
It’s not always about the lifetime argument but rust types and heap vs stack allocation
